log4j-core/src/main/java/org/apache/logging/log4j/core/util/internal/instant/InstantPatternLegacyFormatter.java

```diff
@@ -46,7 +46,10 @@ final class InstantPatternLegacyFormatter implements InstantPatternFormatter {
     private final BiConsumer<StringBuilder, Instant> formatter;
 
     InstantPatternLegacyFormatter(final String pattern, final Locale locale, final TimeZone timeZone) {
-        this.precision = new InstantPatternDynamicFormatter(pattern, locale, timeZone).getPrecision();
+        // Replaces 'n' used in legacy patterns with 'S' to obtain the right precision.
+        // In legacy patterns, the precision of 'n' depends on the pattern length, but
+        // in `DateTimeFormatter`, it is always nanoseconds.
+        this.precision = new InstantPatternDynamicFormatter(pattern.replace('n', 'S'), locale, timeZone).getPrecision();
```

Comment:
Doesn't `IPLF::new`  always get passed the legacy pattern? What does this fix? Is this already an issue in `2.x`, i.e., excluding changes of this PR?

Reply:
Got it. Created #3816 to address this. I will revert this change.